### PR TITLE
WIP: QuickLook Extension for macOS 10.15+

### LINKED ISSUE #3029: Feature Request: support "Quick Look Preview extension" for macOS 10.13 and newer
Quick Look Preview extension is the replacement for the old Quick Look Plug-In (although, this one still works even in macOS 12 Monterey).

In Xcode, it looks like this:
Capture d’écran 2022-04-30 à 20 07 35
It's presented in some WWDC videos like:
https://developer.apple.com/videos/play/wwdc2017/231 for macOS 10.13
https://developer.apple.com/videos/play/wwdc2018/237 for macOS 10.14
https://developer.apple.com/videos/play/wwdc2019/719 for macOS 10.15
The documentation is at:
https://developer.apple.com/documentation/quicklook/qlpreviewingcontroller?language=objc

### PR DESCRIPTION
**Pull request by nevack**
Fixes #3029.

image

### CODE REVIEW

**Comment by nevack** [comment]
@ckerr macOS agents on TeamCity have old SDK installed.
`alpine-musl` GitHub Action job fails in several PRs.

**Comment by Coeur** [comment]
You need to configure the CMakeList.txt to build that target too.

**Comment by nevack** [comment]
> You need to configure the CMakeList.txt to build that target too.

Yep, that's why this is draft.
But CI failures are unrelated to CMake, they must be fixed separately.

**Comment by nevack** [comment]
> For the fun, you could do it in Swift. :D

Yeah, I thought about it as this target is isolated.
But there're some issues to be solved:
1. Compile Swift targets with CMake.
2. Bridges for `libtransmission`.

**Comment by nevack** [comment]
@Coeur What do you think about this implementation in general?

I think I can improve it by using tables, add some fancy icons for file types.

Also it's possible to use HTML based approach like in the old `.qlgenerator` Plugin, but I just don't feel like it's good to do so.

**Comment by Coeur** [comment]
We only need to do a basic implementation (well, that covers the features of the existing QuickLook plugin), and other changes or additions can be done in later pull requests by anyone interested.

**Comment by nevack** [comment]
> We only need to do a basic implementation (well, that covers the features of the existing QuickLook plugin), and other changes or additions can be done in later pull requests by anyone interested.
> 
> 
> 
> You have complete freedom for now.

My current implementation lacks creator info and icons for files, as far as I can tell.

Also trackers and files are listed as text separated by newlines.

Moving to NSTableView is much more involving, but will help to add proper alignment and icons for files.

If creating CMake target for App Extension is not hard, I'll try to merge it before 4.0.0 cut.

This implementation is much more future-proof, looks more natively and supports Dark Appearance out of the box.

**Comment by DevilDimon** [comment]
Can the PR be reviewed in its current state or shall I wait for the final implementation?

**Comment by nevack** [comment]
> Can the PR be reviewed in its current state or shall I wait for the final implementation?

I think it can be reviewed already in this state.

Maybe this review will help me to improve it to a more complete implementation sooner than later

File: macosx/QuickLookExtension/PreviewViewController.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// This file Copyright © 2010-2022 Transmission authors and contributors.
+// It may be used under the MIT (SPDX: MIT) license.
+// License text can be found in the licenses/ folder.
+
+#import <Cocoa/Cocoa.h>
```
Comment (DevilDimon) [inline macosx/QuickLookExtension/PreviewViewController.h:5]: We don't need the whole Cocoa for this, AppKit import would suffice

File: macosx/QuickLookExtension/PreviewViewController.mm [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+// License text can be found in the licenses/ folder.
+
+#import "PreviewViewController.h"
+
+#include <libtransmission/error.h>
+#include <libtransmission/torrent-metainfo.h>
+
+#import "NSStringAdditions.h"
+
+#import <QuickLookUI/QLPreviewingController.h>
```
Comment (DevilDimon) [inline macosx/QuickLookExtension/PreviewViewController.mm:12]:
```suggestion
@import QuickLookUI;
```

File: macosx/QuickLookExtension/PreviewViewController.mm [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+@end
+
+#pragma mark -
+
+@interface PreviewViewController ()<QLPreviewingController>
+
+#pragma mark - IBOutlet
+@property(weak) IBOutlet NSImageView* fileIconImage;
```
Comment (DevilDimon) [inline macosx/QuickLookExtension/PreviewViewController.mm:51]: We keep properties nonatomic across the project. Let's fix it here and in the rest of the file

File: macosx/QuickLookExtension/PreviewViewController.mm [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+#pragma mark - NSViewController
+
+- (NSString*)nibName
+{
+    return @"PreviewViewController";
+}
+
+- (void)loadView
+{
+    [super loadView];
```
Comment (DevilDimon) [inline macosx/QuickLookExtension/PreviewViewController.mm:72]: Is this required? The docs say starting from 10.10 this method and the one above can be omitted as long as the nib name is the same (which it is)

File: macosx/QuickLookExtension/PreviewViewController.mm [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    auto metainfo = tr_torrent_metainfo{};
+    tr_error* error = nullptr;
+    if (!metainfo.parseTorrentFile(url.path.UTF8String, nullptr, &error))
+    {
+        NSError* err = nil;
+        if (error != nullptr)
+        {
+            err = [[TRError alloc] initWith:error];
+        }
+        handler(err);
```
Comment (DevilDimon) [inline macosx/QuickLookExtension/PreviewViewController.mm:89]: Should we return after this line or does the handler terminate the extension automatically?
Either way, I'd make the return explicit

File: macosx/QuickLookExtension/PreviewViewController.mm [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+                [listSection appendFormat:@"%@\n", shortenedFilePathAndSize];
+            }
+
+            self.filesField.hidden = NO;
+            self.filesField.stringValue = listSection;
+        }
+        else
+        {
+            self.fileIconImage.image = [[NSWorkspace sharedWorkspace] iconForFileType:name.pathExtension];
```
Comment (DevilDimon) [inline macosx/QuickLookExtension/PreviewViewController.mm:139]:
```suggestion
            self.fileIconImage.image = [NSWorkspace.sharedWorkspace iconForFileType:name.pathExtension];
```

File: macosx/QuickLookExtension/PreviewViewController.mm [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+#pragma mark - TRError
+
+@interface TRError : NSError
+
+- (instancetype)initWith:(tr_error*)error;
+
+@end
+
+@interface TRError ()
```
Comment (DevilDimon) [inline macosx/QuickLookExtension/PreviewViewController.mm:22]: I am not sure what problem this class solves. We can return the plain NSError supplied with the right parameters. What does an extra class achieve?

**Review by DevilDimon — COMMENTED** [review]
Overall solid work, just a few nits. 

I'll approve after I actually delve deep into the xib when I have time to clone the branch and build from source.
_(it could have been avoided if we used less xibs and did the layout in code though 😄)_

File: macosx/QuickLookExtension/PreviewViewController.mm (hunk 7 above)
Comment (nevack) [reply macosx/QuickLookExtension/PreviewViewController.mm:22]: This was more like a little bit of my playground here.
NSError can be `init`ed with error code and domain, but to provide message (which we already have as `char*` in `tr_error`) you need to subclass.
See [Subclassing notes for NSError](https://developer.apple.com/documentation/foundation/nserror#:~:text=for%20more%20information.-,Subclassing%20Notes,-Applications%20may%20choose)

Yeah, I won't include this in final impl.
Maybe it's good idea to create this NSError<->tr_error wrapper in common place for all macosx related code.

File: macosx/QuickLookExtension/PreviewViewController.mm (hunk 4 above)
Comment (nevack) [reply macosx/QuickLookExtension/PreviewViewController.mm:72]: I'll test without it and remove if really it is not needed anymore.

**Comment by nevack** [comment]
> Overall solid work, just a few nits.
> 
> I'll approve after I actually delve deep into the xib when I have time to clone the branch and build from source. _(it could have been avoided if we used less xibs and did the layout in code though 😄)_

I think I can try to reimplement it with code.
It was just faster to create initial PoC with XIB :wink:

File: macosx/QuickLookExtension/PreviewViewController.mm (hunk 7 above)
Comment (DevilDimon) [reply macosx/QuickLookExtension/PreviewViewController.mm:22]: You don't actually need to subclass though. Localized description is stored in the user info dictionary and can be retrieved from any NSError. See [localizedDescription docs](https://developer.apple.com/documentation/foundation/nserror/1414418-localizeddescription?language=objc)

Comment (nevack) [reply macosx/QuickLookExtension/PreviewViewController.mm:22]: Ohh, thanks!

**Comment by DevilDimon** [comment]
Regarding the feasibility of building this with Swift: there is some Swift-C++ [interop support](https://github.com/apple/swift/blob/main/docs/CppInteroperability/CppInteroperabilityStatus.md), some of it surprisingly good. 
It requires some CMake flags though and I am not aware how backwards-compatible that is and whether the imported feature set is enough for this PR. You could try playing with that if you're interested, it looks like it's possible without any manual bindings.

**Comment by nevack** [comment]
Rebase, Code review fixes, Removed my signing info.

**Comment by nevack** [comment]
@ckerr FYI, these PR is triggering `alpine` and `debian` actions.
I think changes in `Transmission.xcodeproj` should be ignored for these runners.

**Comment by nevack** [comment]
@DevilDimon Can you help here?
```
/Users/teamcity/buildAgent1/work/1f3bff43dd55cfe9/macosx/QuickLookExtension/PreviewViewController.mm:12:9: fatal error: module 'QuickLookUI' not found
@import QuickLookUI;
 ~~~~~~~^~~~~~~~~~~
```

No problem locally, I think Xcode/SDK is outdated on teamcity agent.

File: macosx/QuickLookExtension/PreviewViewController.h [hunk 8]
```diff
@@ -0,0 +1,9 @@
+// This file Copyright © 2010-2022 Transmission authors and contributors.
+// It may be used under the MIT (SPDX: MIT) license.
+// License text can be found in the licenses/ folder.
+
+@import AppKit;
```
Comment (DevilDimon) [inline macosx/QuickLookExtension/PreviewViewController.h:5]: This is probably the reason why `@import QuickLookUI` doesn't compile. Objective-C++ disallows `@import` statements in header files (save for a few select cases) and the errors when doing so are cryptic. I'd change this import to a preprocessor statement to see if it compiles then.

**Comment by nevack** [comment]
@DevilDimon Can you help me here with creating CMake subproject for appex?
I'm not really good with CMake packaging for macOS, sorry 😞

**Comment by DevilDimon** [comment]
Yeah, I'll try my hand at that soon.

**Comment by nevack** [comment]
Closing in favor of #7213

**Pull request closed without merging**